### DISCUSSION TITLE
Wrong camera for best focus abilities

### DISCUSSION THREAD

**Discussion by robertnilssongreenconse** [mebjas/html5-qrcode#655] [General]
First of: great work with the code reader!

I got a question though... realised after some testing that some phones are better and some are worse at reading barcodes; my Android was one of them that had problems. I did realise that the camera was not that good at focusing although when I used native apps the image was crystal clear.

After forcing the code to use, not the first camera returned by the "facingMode: "environment" -setting, but rather pick the other one by id (my phone has 2 backward facing cameras), the image was much better, close to crisp.
So, is there any way of picking the camera with the best focal ability? There is a constraint in the specification from Mozilla with the name "focusDistance". Can this be used in any way to pick the camera thats best suitet for close range scanning?

**Comment by andredebarrospinto** [comment]
We had the same issue with small barcodes and especially with the most recent cameras with 3 lenses like Iphones, I already tested with Samsung and it is working better.
 
I think if we can zoom or select the area to be scanned and focus should work. And it should be automatically.

Anyway to implement that?

**Reply by robertnilssongreenconse** [reply]
I managed to improve barcode-scanning performance for mostly android devices by NOT using facingMode: "environment" bcause it gives you a shitty camera most of the times. What I did was I took the list of devices and sorted them by the Label. Then I took the camera with index 0 which mostly represents the camera that the manufacturer inteded for default use when starting camera app on the phone. The label is "camera2 0, facing back". The 0 is a give away! After using this camera the scanner box flashed like a stoboscope and I could even scan a dirty barcode from some distance and even when it was tilted!

However... the iphone does not care about this. My debug-possibilities for iphone is limited and I cant get it to give away the list of cameras so I can try picking another one....
